### DISCUSSION TITLE
[Question] understandings of eKuiper

### DISCUSSION THREAD

**Discussion by guowei-gong** [lf-edge/ekuiper#1402] [General]
Hi，这是我对 `ekuiper` 的一些理解，说的不对的地方请**斧正**。感谢。

1. 没有 `ekuiper` 时，数据规则放在 `EMQX` 云端进行，此时流量实际已经打到了云端。
有 `ekuiper` 时，`ekuiper` 运行在设备端，也就是边缘端，在边缘端定义了数据规则，将过滤后的数据打到 `EMQX`，节约了带宽、流量。
2. `ekupier` 创建的是流。发送到主题的数据，会被这个流订阅并处理。
3. 定义流，其实就像关系型数据库中定义一张表。
4. `ekuiper` 的加工数据可以通过 `Go` 进行扩展。

**Comment by ngjaying** [comment]
这几点都是基于处理 MQTT 数据并与 EMQ 协同，没有问题。不过，eKuiper 本身并不依附 EMQX 或者 MQTT，流的类型可以是别的协议，甚至是扩展私有协议。扩展的部分处理数据处理函数之外，还可以是数据连接的部分包括输入的source和输出的sink。

**Reply by guowei-gong** [reply]
感谢回复，这让我对于 ekuiper 的数据源、数据输出、加工数据都可以通过 Go 进行扩展有了更深的理解。我正在尝试通过 EMQX 在本地搭建一个 MQTT 消息服务器，ekuiper 将创建流，使用我本地的消息服务器，并通过 MQTT 客户端工具发送测试数据，最终将过滤后的数据通过文件的形式保存。顺便一提，ekuiper 通过规则过滤的数据输出，与 EMQX 接收到的数据，有什么区别？我认为， EMQX 吐数据，它接收的是更多设备的数据，像一个汇总。而 ekuiper 仅仅是这台边缘设备过滤后的数据吗？再次感谢你的回复。

**Reply by ngjaying** [reply]
数据内容/格式是业务定义的。eKuiper 本身也可以接收来自多个设备的输入，只不过当前是单用例而非集群运行，处理能力不如 EMQX 集群。

**Reply by guowei-gong** [reply]
tks u reply! happy coding!

**Reply by guowei-gong** [reply]
打扰一下，我有一个关于 ekuiper + emqx 协同，数据清洗的相关问题。

ekuiper 的工作方式是创建流、定义规则、输出数据。数据的来源可以是 mqtt 等，目前我使用的是 mqtt 协议发送的数据，ekuiper 的流，它的源数据，是通过代理的方式获取，例如 ekuiper_1 订阅了发送到 emqx_1 的 mqtt 协议的数据，这个数据发送到 topic_1，那么此时，我的预期是 ekuiper_1 会清洗发送到 emqx_1 的 mqtt 协议的数据，并将清洗后的数据发送到 emqx_1 的 topic_1 主题，然而，事实是数据的流量已经打到了 emqx_1，当然，ekuiper_1 也收到了该数据并根据规则输出了过滤后的数据。

所以我对于这个协同使用，没有很理解，ekuiper 的工作是要在启动时指定消息服务器的，可是指定的消息服务器，如果和过滤数据输出的目标消息服务器是一致的，那么不是与 清洗数据 后发送到目标服务器不符了吗？

因为流量已经打到了消息服务器。

**Reply by ngjaying** [reply]
我也不太懂你的问题，你是说处理后的数据没有发回 topic_1 吗？这样发回原处会导致无线循环。数据处理后的动作是由规则里的actions指定的。比如下面的规则，结果会写到log里，另外发到 mqtt 的 result_1 主题。

```json
{
  "id": "ruleLookup",
  "sql": "SELECT demo.id, rms(signal) FROM demo INNER JOIN sqlTable ON demo.id = sqlTable.device_id AND trigger = TRUE",
  "actions": [{
    "log": {
    },
    "mqtt": {
      "server": "tcp://mybroker:1883",
      "topic": "result_1",
      "sendSingle": true
    }
  }]
}
```

**Reply by guowei-gong** [reply]
抱歉，我的描述是糟糕的😰，我画了一张图辅助我的问题，希望能更清晰的描述它，图片如下。
![image](https://user-images.githubusercontent.com/62499904/190547953-8a329917-5258-4ada-8c76-b1fd3bbec2dc.png)

我的预期： ekuiperA 过滤来自数据源的数据，ekuiperA 将过滤后的数据发送到 emqxA。

事实：数据源发送的数据先打到了 emqxA。当然，ekuiperA 也会将根据规则过滤后数据发送到 emqxA。emqxA 会收到两份数据！

**Reply by guowei-gong** [reply]
顺便一提，我希望在图中绿色线条的部分完成数据过滤，而不是 emqxA 会收到未过滤的数据。

**Reply by ngjaying** [reply]
EMQX作为MQTT broker，一定会收到所有走MQTT的数据，它是通过 topic 来区分数据的。我觉得你这里可能需要关注的是 topic 的配置。未过滤的数据发到一个 topic , 例如 `input`； eKuiper stream 的 dataSource 配置成 `input`，处理完之后sink配置另一个 topic `output`. 这样 input里面都是待处理的数据，output里面都是处理好的数据。

**Reply by ngjaying** [reply]
感觉你想要的是接入本地的数据源，而不是云端的 EMQX。本地其实也有轻量化的MQTT broker 比如 NanoMQ，或者如果本地的数据本身是别的协议产生的，也可以用 eKuiper 直接接入。

欢迎加入微信群讨论，下面有链接：
https://github.com/lf-edge/ekuiper/tree/master/docs/en_US#get-help

**Reply by guowei-gong** [reply]
感谢您的回复，我对于 emqx 的理解是差劲的（或者说我对`消息服务器`这个词，忽略的最重要：`消息`，这两个字），emqx 就是一个**代理**。目前我描绘的场景是这样的。如下图。
![image](https://user-images.githubusercontent.com/62499904/190552678-60fb8edb-8ebd-4957-8f8f-5b248936ce8b.png)

我的脑海里有一个想法，如下图（这样，EMQX 就不会收到两次数据了，不过这样好像失去了使用 EMQX 作为消息服务器的意义了，我是一个新手，抱歉）。我已经理解了您的回复。
![image](https://user-images.githubusercontent.com/62499904/190553712-0c47f265-c2b4-4b75-9aa0-e92abbde4ec5.png)

**Reply by guowei-gong** [reply]
> 感觉你想要的是接入本地的数据源，而不是云端的 EMQX。本地其实也有轻量化的MQTT broker 比如 NanoMQ，或者如果本地的数据本身是别的协议产生的，也可以用 eKuiper 直接接入。
> 
> 欢迎加入微信群讨论，下面有链接： https://github.com/lf-edge/ekuiper/tree/master/docs/en_US#get-help

已申请 :-)

**Reply by guowei-gong** [reply]
> 感觉你想要的是接入本地的数据源，而不是云端的 EMQX。本地其实也有轻量化的MQTT broker 比如 NanoMQ，或者如果本地的数据本身是别的协议产生的，也可以用 eKuiper 直接接入。
> 
> 欢迎加入微信群讨论，下面有链接： https://github.com/lf-edge/ekuiper/tree/master/docs/en_US#get-help

amazing! 是的，我好像没有正确的理解我的需求，不过我的关注点一直在 `source`。

**Reply by guowei-gong** [reply]
按照您的回复，目前 `EMQX` 被看做是一个数据源吗？如果这样，我发起这篇讨论的观点好像出现了错误，观点如下：
1. 没有 ekuiper 时，数据规则放在 EMQX 云端进行，此时流量实际已经打到了云端。
有 ekuiper 时，ekuiper 运行在设备端，也就是边缘端，在边缘端定义了数据规则，将过滤后的数据打到 EMQX，节约了带宽、流量。

此时的数据都会打到 EMQX，何来引入了 ekuiper 反而节约了带宽、流量这个说法呢？

**Reply by ngjaying** [reply]
1. EMQX是一个MQTT broker，是MQTT协议的消息汇聚。 eKuiper 这边只是有 MQTT source， MQTT sink 跟别的 source 和 sink都是并列的关系，也不关心具体的MQTT broker 实现。
2. 物联网场景中，边缘端通常由各种设备/传感器等产生数据。这些数据量特别大，单独看又不是特别重要。eKuiper的作用就是在边缘端就近对这些数据进行处理，过滤啊，转换啊，聚合啊，这样只提取出有用的数据传到云端，相比全量上传就是节约了带宽。这边假设的是数据源和 eKuiper 都部署在边缘端，数据到 eKuiper 不需要经过外网，没有带宽消耗。

**Reply by ngjaying** [reply]
一般 eKuiper source端接的是边缘的数据，sink端发到边缘或云端。如果数据本身就来自云端，把eKuiper部署到同样的云端位置就行了

**Reply by guowei-gong** [reply]
感谢您的帮助 :-)
